### ISSUE TITLE
user validation

### ISSUE THREAD

**Issue by amaurydar** [KxSystems/jupyterq#26]
Is it possible to restrict access to the processes that have opened a port, just like with command-line argument `-U password_file.txt` ?
(do both the server and the kernel open a port ?)

**Comment by jhanna-kx** [comment]
This isn't possible at the moment.

Both processes need to open ports, and make connections to each other.

We can set command line arguments for either process currently, see https://code.kx.com/q/ml/jupyterq/command-line/

However since both kernel and server need to open connections to each other and there will be no way from the jupyter notebook server or console to specify the username/password we'd need a change to specify a username and password for the kernel<->server connections (from an environment variable probably).

**Comment by amaurydar** [comment]
Yes, setting a username and password in an environment variable would be very useful, as currently it is very unsafe to have processes with open ports on shared machines.

**Comment by jhanna-kx** [comment]
The changes in #28 close the ports the kernel and server processes useafter connections are establised. 

Also in case you subsequently want to open a port for the server or kernel process but want user authentication enabled, you can now specify -U (not -u) as command line options for the server and kernel, if doing this you need to provide valid a username and password to allow connections to the kernel and server in an environment variable `JUPYTERQ_LOGIN` e.g. kernel.json could look like this

```json
{
 "argv": [
  "q",
  "jupyterq_kernel.q",
  "-U",
  "kernuser.txt",
  "-cds",
  "{connection_file}"
 ],
 "display_name": "Q (kdb+)",
 "language": "q",
 "env": {"JUPYTERQ_SERVERARGS":"-U servuser.txt","MPLBACKEND":"Agg"}
}
```
And then you can use separate password files for the kernel and server, but `JUPYTERQ_LOGIN` must contain a username password combination valid for **both** processes.
```bash
$ cat servuser.txt # password stored hashed
user1:5f4dcc3b5aa765d61d8327deb882cf99
user2:5f4dcc3b5aa765d61d8327deb882cf99
$ cat kernuser.txt # passwords can also be stored in plaintext
user1:password
$ export JUPYTERQ_LOGIN=user1:password
#should work
$ jupyter-console --kernel=qpk
....
Jupyter console 5.1.0

KDB+ v3.6 2018.06.01 kdb+ kernel for jupyter, jupyterQ vdevelopment
In [1]: 

...
#shouldn't work as login not valid for both kernel and server 
$ JUPYTERQ_LOGIN=user2:password jupyter-console --kernel=qpk processes
*******************************************
** Wrong user:password in JUPYTERQ_LOGIN **
*******************************************
Press Ctrl+C
```

At the moment this is only in the latest version of the code, We'll update the docs and push new releases including it shortly.
